### PR TITLE
added percent_change function

### PR DESCRIPTION
**Pull request by shahsaurabh0605**
I have added percent_change function in vector.rb. Have a look and suggest me of any changes to be done.

### CODE REVIEW

File: lib/daru/vector.rb [hunk 1]
```diff
@@ -576,6 +576,38 @@ def sorted_data &block
       @data.to_a.sort(&block)
     end
 
+    # Create a new vector by specifying the periods of nils.
+    #
+    # == Description
```
Comment (v0dro) [inline lib/daru/vector.rb]: Follow YARD documentation syntax.

**Comment by v0dro** [comment]
Could you add a test case?

**Comment by shahsaurabh0605** [comment]
Added the specs and rewritten documentation in YARD syntax.

**Comment by v0dro** [comment]
I didn't see this before, but this is a stats function and belongs in this file: https://github.com/v0dro/daru/blob/master/lib/daru/maths/statistics/vector.rb

**Comment by shahsaurabh0605** [comment]
Sorry, I missed that. Added the function in appropriate place.

File: lib/daru/maths/statistics/vector.rb [hunk 2]
```diff
@@ -377,6 +377,36 @@ def sample_without_replacement(sample=1)
...
+        #   #   a	
+        #   #   b	   -0.3333333333333333
+        #   #   c	   -0.5
+        #   #   f	   -0.5
+        def percent_change periods = 1
+          value = []
+          arr = []
+          new_index = [] 
+          self.index.each do |e|
+            if self[e].is_a?(Numeric)
```
Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: You can replace this loop with `self.type == :numeric`. When the type is numeric it indicates that all the members are numerics or missing. You can then call only_valid to get only the numerical values.

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: And btw, percent change for any data apart from numeric should not be supported.

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: So if a string is encountered, it's supposed to raise an error.

**Comment by shahsaurabh0605** [comment]
Now, the percent change is applicable to only numeric vectors.

File: lib/daru/maths/statistics/vector.rb [hunk 3]
```diff
@@ -377,6 +377,35 @@ def sample_without_replacement(sample=1)
...
+        #   #=>
+        #   #   <Daru::Vector:28713060 @name = nil @size: 5 >
+        #   #              nil
+        #   #   a	
+        #   #   f	   0.5
+        #   #   t	   0.0
+        #   #   i	   0.3333333333333333
+        #   #   k          0.25
+        def percent_change periods = 1
+          value = []
```
Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: remove this line

File: lib/daru/maths/statistics/vector.rb [hunk 4]
```diff
@@ -377,6 +377,35 @@ def sample_without_replacement(sample=1)
...
+          arr = []
```
Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: put this inside the `if`.

File: spec/math/statistics/vector_spec.rb [hunk 5]
```diff
@@ -383,6 +383,14 @@
       expect(acf[3]).to be_within(0.001) .of(0.486)
     end
   end
+  
+  context "#percent_change" do
+    it "calculates percent change" do
+      vector = Daru::Vector.new([4,6,6,8,10],index: ['a','f','t','i','k'])
```
Comment (v0dro) [inline spec/math/statistics/vector_spec.rb]: What if there's missing data in the vector? Vectors with all numeric data and missing data still qualify as numeric vectors.

**Comment by shahsaurabh0605** [comment]
I have made the changes such that `percent_change` now handles only numeric vectors alongwith missing data.

File: lib/daru/maths/statistics/vector.rb [hunk 6]
```diff
@@ -377,6 +377,39 @@ def sample_without_replacement(sample=1)
...
+        #   #   i	   0.3333333333333333
+        #   #   k          0.25
+        def percent_change periods = 1
+          type == :numeric or raise TypeError, "Vector must be numeric"
+          arr = []
+          i = periods
+          (periods...size).each do |j|
+            if @missing_values.has_key?(@data[i])
+              arr[j] = nil
+              @data.delete_at(i)
```
Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: Didn't you notice that `delete_at` changes the data of the vector on which percent_change is being called?

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: You know you really should think about all the side effects your code can have before you commit.

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: Why not simply call `Vector#only_valid` and work on the vector that is returned by it?

**Comment by shahsaurabh0605** [comment]
I have made the changes such that `Vector#only_valid` is used. I will try to be more careful of the side effects from now on.

File: spec/math/statistics/vector_spec.rb (hunk 5 above)
Comment (v0dro) [inline spec/math/statistics/vector_spec.rb]: Could you add a test case for missing data?

Comment (v0dro) [inline spec/math/statistics/vector_spec.rb]: Here's a case:
``` ruby

 v=Daru::Vector.new([nil,6,nil,8,10],index: ['a','f','t','i','k'])
```

File: lib/daru/maths/statistics/vector.rb [hunk 7]
```diff
@@ -377,6 +377,40 @@ def sample_without_replacement(sample=1)
...
+        #   #   <Daru::Vector:28713060 @name = nil @size: 5 >
+        #   #              nil
+        #   #   a	
+        #   #   f	   0.5
+        #   #   t	   0.0
+        #   #   i	   0.3333333333333333
+        #   #   k          0.25
+        def percent_change periods = 1
+          type == :numeric or raise TypeError, "Vector must be numeric"
+          value = self.only_valid
```
Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: Now that I think of it, you can get away with working with `@data` only.

Here's a sample:
``` ruby

def percent_change periods = 1
   type == :numeric or raise TypeError, "Vector must be numeric"
   arr = []
   i = 1 
   ind = @data.find_index{|x|!x.nil?}
   (periods...size).each do |j|
     if j==ind || @missing_values.has_key?(@data[j])
       arr[j] = nil
     else
       arr[j] = (@data[i] - @data[i - 1]) / @data[i - 1].to_f
       i+=1
     end 
   end
   Daru::Vector.new(arr, index: @index, name: @name)
 end
```

Comment (shahsaurabh0605) [inline lib/daru/maths/statistics/vector.rb]: This code won't work for test cases like this : 
`vector = Daru::Vector.new([2,4,5,nil,6,7,nil])`
Actually, counter `i` must run through only numbers excluding the nils. So working with `only_valid` will be the best in this case. I initially tried to do this with deleting nils from `@data` but that is obviously not the way out.

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: Well then can you please make it work? Refer the pandas `pct_change` function.

Comment (shahsaurabh0605) [inline lib/daru/maths/statistics/vector.rb]: Is there any problem with using `only_valid`? I think this should be the best to serve our purpose here.

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: I don't have a problem. But will it work with missing data?

Comment (shahsaurabh0605) [inline lib/daru/maths/statistics/vector.rb]: I tested it for all the test cases including missing data. According to me it works!

Comment (v0dro) [inline lib/daru/maths/statistics/vector.rb]: OK then can you add those tests to the spec?

**Comment by shahsaurabh0605** [comment]
Added the tests.

**Pull request merged**